File: vreader/Services/AI/ChapterTranslationService.swift
// Purpose: Translates one chapter unit for feature #56 bilingual reading.
// Pipeline: cache lookup → (on miss) segment → chunk → one AIService request
// per chunk → strict JSON-array decode → per-segment fallback on any decode /
// count / element mismatch → recombine → cache-write.
//
// Key decisions:
// - A SEPARATE actor (not a `PersistenceActor`/`AIService` extension) — it
//   composes the WI-2 store + WI-4 chunker/contract + WI-5 config seam.
// - The AI side is reached through the `TranslationRequestSending` protocol
//   (`AIService` conforms) so tests inject a deterministic mock — the
//   `LibraryPersisting`/`BookImporting` boundary-protocol precedent.
// - On a cache HIT the cached segments are returned with `fromCache == true`
//   and ZERO API calls (acceptance criterion (c)).
// - On any chunk decode failure (`TranslationChunkContract.DecodeError`) the
//   service falls back to one-segment-per-request under the SAME config +
//   style (v4 Gate-2 F5).
// - `Task.checkCancellation()` runs between chunks so a cancelled prefetch /
//   global job stops promptly (edge case (b)).
//
// @coordinates-with: ChapterTranslationStore.swift, ChapterSegmenter.swift,
//   ChapterTranslationChunker.swift, TranslationChunkContract.swift,
//   ResolvedAIProviderConfig.swift, AIService.swift,
//   dev-docs/plans/20260519-feature-56-bilingual-reading.md (WI-6)

import Foundation
import OSLog

/// The AI-request boundary the translation service depends on — `AIService`
/// conforms in production; tests inject a deterministic mock.
protocol TranslationRequestSending: Sendable {
    /// Sends one translation request through a pre-resolved provider config.
    func sendTranslationRequest(
        _ request: AIRequest,
        using config: ResolvedAIProviderConfig
    ) async throws -> AIResponse
}

extension AIService: TranslationRequestSending {
    func sendTranslationRequest(
        _ request: AIRequest,
        using config: ResolvedAIProviderConfig
    ) async throws -> AIResponse {
        try await sendRequest(request, using: config)
    }
}

/// A typed failure from chapter translation.
enum ChapterTranslationError: Error, Equatable {
    /// The device is offline and the unit is not cached.
    case offline
    /// The provider call failed (network / API error).
    case providerFailed(String)
    /// The translation was cancelled.
    case cancelled
}

/// The outcome of translating one unit.
struct ChapterTranslationResult: Sendable, Equatable {
    /// One translated segment per source segment, in order.
    let segments: [String]
    /// `true` when served entirely from the disk cache (no API call).
    let fromCache: Bool
}

/// Granularity of translation segmentation (mirrors `PerBookSettings`'
/// `bilingualGranularity`).
enum TranslationGranularity: String, Sendable {
    case paragraph
    case sentence
}

/// Actor translating one chapter unit, with a provider-aware disk cache.
actor ChapterTranslationService {

    /// The per-chunk character budget. Conservative — well under any
    /// mainstream provider's context window so a chunk + its prompt scaffold
    /// fit comfortably; large chapters are split into several requests.
    static let defaultMaxCharsPerChunk = 6000

    private let sender: any TranslationRequestSending
    private let store: ChapterTranslationStore
    private let promptVersion: String
    private let maxCharsPerChunk: Int
    private let log = Logger(subsystem: "com.vreader.app", category: "ChapterTranslationService")

    init(
        sender: any TranslationRequestSending,
        store: ChapterTranslationStore,
        promptVersion: String,
        maxCharsPerChunk: Int = ChapterTranslationService.defaultMaxCharsPerChunk
    ) {
        self.sender = sender
        self.store = store
        self.promptVersion = promptVersion
        self.maxCharsPerChunk = maxCharsPerChunk
    }

    /// Bug #306: a CACHE-ONLY lookup that needs NO provider config. Returns the
    /// cached translation when a fresh (count-matching) row exists, else nil.
    /// Lets the prefetcher serve an already-translated chapter BEFORE the
    /// provider gate (`resolveProviderConfig`), so a cached chapter still renders
    /// when AI is later disabled / unconfigured / key-less — previously the gate
    /// threw first and the disk cache (inside `translate`) was never reached.
    func cachedTranslation(
        bookFingerprintKey: String,
        unit: TranslationUnitID,
        sourceText: String,
        targetLanguage: String,
        providerProfileID: UUID,
        granularity: TranslationGranularity = .paragraph
    ) async -> ChapterTranslationResult? {
        let lookupKey = ChapterTranslationRecord.lookupKey(
            bookFingerprintKey: bookFingerprintKey,
            unitStorageKey: unit.storageKey,
            targetLanguage: targetLanguage,
            providerProfileID: providerProfileID,
            promptVersion: promptVersion)
        let segments: [String]
        switch granularity {
        case .paragraph: segments = ChapterSegmenter.paragraphs(in: sourceText)
        case .sentence:  segments = ChapterSegmenter.sentences(in: sourceText)
        }
        guard let cached = await store.translation(forKey: lookupKey),
              cached.sourceParagraphCount == segments.count else { return nil }
        return ChapterTranslationResult(segments: cached.translatedSegments, fromCache: true)
    }

    /// Translates `unit`'s source text into `targetLanguage`. Serves from the
    /// disk cache on a hit; on a miss segments → chunks → requests → decodes →
    /// caches. Throws `ChapterTranslationError` on a provider failure or
    /// cancellation.
    func translate(
        bookFingerprintKey: String,
        unit: TranslationUnitID,
        sourceText: String,
        targetLanguage: String,
        providerProfileID: UUID,
        config: ResolvedAIProviderConfig,
        style: TranslationStyle,
        granularity: TranslationGranularity = .paragraph,
        // Bug #311: optional real progress source. Fired after each chunk
        // completes with `(completedChunks, totalChunks)` so a caller (the
        // re-translate VM) can drive an honest N-of-M progress bar instead of
        // a faked 0.5 pin during the opaque per-chunk network phase. nil by
        // default — the whole-book coordinator + bilingual paths don't use it.
        onChunkProgress: (@Sendable (Int, Int) -> Void)? = nil
    ) async throws -> ChapterTranslationResult {
        let lookupKey = ChapterTranslationRecord.lookupKey(
            bookFingerprintKey: bookFingerprintKey,
            unitStorageKey: unit.storageKey,
            targetLanguage: targetLanguage,
            providerProfileID: providerProfileID,
            promptVersion: promptVersion)

        // Segment the source per the requested granularity FIRST — the
        // segment count is needed to detect a stale cache row.
        let segments: [String]
        switch granularity {
        case .paragraph: segments = ChapterSegmenter.paragraphs(in: sourceText)
        case .sentence:  segments = ChapterSegmenter.sentences(in: sourceText)
        }

        // Cache lookup. A row is served only when its `sourceParagraphCount`
        // still matches the live chapter — a chapter whose source has since
        // changed (content-replacement rule edit, re-import) produces a
        // mismatch, which is treated as STALE: the row is dropped and the
        // chapter re-translated (plan audit-driven addition).
        if let cached = await store.translation(forKey: lookupKey) {
            if cached.sourceParagraphCount == segments.count {
                return ChapterTranslationResult(
                    segments: cached.translatedSegments, fromCache: true)
            }
            log.info("Stale cache row (count \(cached.sourceParagraphCount) != live \(segments.count)); re-translating")
            // A delete failure does not block re-translation — the later
            // upsert refreshes the same lookupKey regardless. Logged so the
            // swallow is visible (rule 50 §6).
            do {
                try await store.deleteTranslation(forKey: lookupKey)
            } catch {
                log.error("Stale-row delete failed (upsert will still refresh it): \(String(describing: error), privacy: .public)")
            }
        }

        guard !segments.isEmpty else {
            return ChapterTranslationResult(segments: [], fromCache: false)
        }

        // Chunk → translate each chunk → recombine in source order.
        let chunks = ChapterTranslationChunker.chunk(
            segments: segments, maxCharsPerChunk: maxCharsPerChunk)
        var translated = [String](repeating: "", count: segments.count)

        // Bug #330: a single chunk's provider failure must NOT abort the whole
        // chapter — leave that chunk's segments source-only (empty translation)
        // and continue. But if EVERY chunk fails (a genuine outage, not just one
        // over-budget chunk), surface the error rather than silently returning an
        // all-source-only result.
        var completedChunks = 0
        var anyChunkSucceeded = false
        var lastChunkError: Error?
        for chunk in chunks {
            do {
                try Task.checkCancellation()
            } catch {
                throw ChapterTranslationError.cancelled
            }
            let chunkSegments = chunk.map { segments[$0] }
            do {
                let chunkResult = try await translateChunk(
                    chunkSegments, targetLanguage: targetLanguage, config: config, style: style)
                for (offset, segmentIndex) in chunk.enumerated() {
                    translated[segmentIndex] = chunkResult[offset]
                }
                anyChunkSucceeded = true
            } catch is CancellationError {
                // The between-chunk / per-piece Task.checkCancellation() throws a
                // raw CancellationError — surface it as the typed .cancelled.
                throw ChapterTranslationError.cancelled
            } catch ChapterTranslationError.cancelled {
                // Bug #330 (Codex): `send()` maps a provider-side cancellation to
                // the TYPED `.cancelled`. It must NOT be degraded-and-continued —
                // a cancel aborts the whole translation. Rethrow before the
                // generic degradation catch below.
                throw ChapterTranslationError.cancelled
            } catch {
                // Bug #330: degrade this chunk to source-only and keep going.
                log.error("Chunk failed (segments \(chunk) left source-only): \(String(describing: error), privacy: .public)")
                lastChunkError = error
            }
            // Bug #311: real N-of-M progress — fire AFTER the chunk resolves
            // (success or graceful-degrade) so the count reflects committed work.
            completedChunks += 1
            onChunkProgress?(completedChunks, chunks.count)
        }
        if !anyChunkSucceeded, let err = lastChunkError {
            throw err
        }

        // Cache-write the recombined ordered translation. A store-write
        // failure does not fail the translation — the caller still gets the
        // freshly translated text; it just won't be cached this time. The
        // swallow is logged so the failure is visible (rule 50 §6).
        //
        // Bug #330: do NOT cache a PARTIALLY-degraded result. If a chunk failed
        // (its segments are source-only ""), caching would make the gap permanent
        // — a later read would hit the cache and never retry the failed chunk.
        // Skip the write so a re-read re-attempts the failed segments; the caller
        // still gets this run's partial result.
        if lastChunkError == nil {
            do {
                try await store.upsert(ChapterTranslationRecord(
                    bookFingerprintKey: bookFingerprintKey,
                    unitStorageKey: unit.storageKey,
                    targetLanguage: targetLanguage,
                    providerProfileID: providerProfileID,
                    promptVersion: promptVersion,
                    translatedSegments: translated,
                    sourceParagraphCount: segments.count))
            } catch {
                log.error("Cache-write failed (translation still returned): \(String(describing: error), privacy: .public)")
            }
        }

        return ChapterTranslationResult(segments: translated, fromCache: false)
    }

    /// Bug #268: translates a PRE-SEGMENTED list of source segments directly,
    /// bypassing `ChapterSegmenter` AND the disk cache. Used by the bilingual
    /// EPUB divergence-fallback: when the DOM leaf-enumerate's block count
    /// diverges from the plain-text paragraph segmentation (nested `<pre>` /
    /// mixed-content `<blockquote>`), translating the enumerate's OWN block
    /// `text[]` makes blocks↔segments 1:1 BY CONSTRUCTION — eliminating the
    /// whole-chapter source-only fallback. The returned array is always the same
    /// length as `segments`. No disk cache: the fallback is rare, and the
    /// plain-text path's cache row is keyed by a different segment count that
    /// must not be thrashed.
    func translatePreSegmented(
        segments: [String],
        targetLanguage: String,
        config: ResolvedAIProviderConfig,
        style: TranslationStyle
    ) async throws -> [String] {
        guard !segments.isEmpty else { return [] }
        let chunks = ChapterTranslationChunker.chunk(
            segments: segments, maxCharsPerChunk: maxCharsPerChunk)
        var translated = [String](repeating: "", count: segments.count)
        // Bug #330: same graceful degradation as `translate` — a single chunk's
        // failure leaves its segments source-only; an all-chunks failure surfaces
        // the error.
        var anyChunkSucceeded = false
        var lastChunkError: Error?
        for chunk in chunks {
            do {
                try Task.checkCancellation()
            } catch {
                throw ChapterTranslationError.cancelled
            }
            let chunkSegments = chunk.map { segments[$0] }
            do {
                let chunkResult = try await translateChunk(
                    chunkSegments, targetLanguage: targetLanguage, config: config, style: style)
                for (offset, segmentIndex) in chunk.enumerated() {
                    translated[segmentIndex] = chunkResult[offset]
                }
                anyChunkSucceeded = true
            } catch is CancellationError {
                throw ChapterTranslationError.cancelled
            } catch ChapterTranslationError.cancelled {
                // Bug #330 (Codex): typed cancellation from `send()` must abort,
                // not degrade.
                throw ChapterTranslationError.cancelled
            } catch {
                log.error("Pre-segmented chunk failed (segments \(chunk) source-only): \(String(describing: error), privacy: .public)")
                lastChunkError = error
            }
        }
        if !anyChunkSucceeded, let err = lastChunkError {
            throw err
        }
        return translated
    }

    // MARK: - Private

    /// Translates one chunk: a single whole-chunk request with a strict
    /// JSON-array decode; on any decode / count / element mismatch, falls back
    /// to one request per segment (still under the same config + style).
    private func translateChunk(
        _ chunkSegments: [String],
        targetLanguage: String,
        config: ResolvedAIProviderConfig,
        style: TranslationStyle
    ) async throws -> [String] {
        // Bug #330: an oversized SINGLE segment (the chunker gives it its own
        // chunk but does NOT sub-split it) is sub-split into ≤budget pieces,
        // translated piece by piece, and rejoined into the one segment's
        // translation — sending the whole oversized paragraph would overflow the
        // provider's context window → error.
        if chunkSegments.count == 1 {
            let pieces = ChapterTranslationChunker.subSplit(
                chunkSegments[0], maxChars: maxCharsPerChunk)
            if pieces.count > 1 {
                var joined = ""
                for piece in pieces {
                    try Task.checkCancellation()
                    let piecePrompt = TranslationChunkContract.userPrompt(
                        segments: [piece], targetLanguage: targetLanguage, style: style)
                    let pieceResponse = try await send(prompt: piecePrompt, config: config)
                    if let decoded = try? TranslationChunkContract.decode(
                        pieceResponse.content, expectedCount: 1) {
                        joined += decoded[0]
                    } else {
                        joined += pieceResponse.content.trimmingCharacters(in: .whitespacesAndNewlines)
                    }
                }
                return [joined]
            }
        }

        // First attempt — the whole chunk in one request.
        let prompt = TranslationChunkContract.userPrompt(
            segments: chunkSegments, targetLanguage: targetLanguage, style: style)
        let response = try await send(prompt: prompt, config: config)
        if let decoded = try? TranslationChunkContract.decode(
            response.content, expectedCount: chunkSegments.count) {
            return decoded
        }

        // Decode failed → per-segment fallback.
        log.warning("Chunk decode failed; falling back to per-segment requests")
        var perSegment: [String] = []
        perSegment.reserveCapacity(chunkSegments.count)
        for segment in chunkSegments {
            try Task.checkCancellation()
            let onePrompt = TranslationChunkContract.userPrompt(
                segments: [segment], targetLanguage: targetLanguage, style: style)
            let oneResponse = try await send(prompt: onePrompt, config: config)
            if let oneDecoded = try? TranslationChunkContract.decode(
                oneResponse.content, expectedCount: 1) {
                perSegment.append(oneDecoded[0])
            } else {
                // Even the per-segment decode failed — keep order intact with
                // the raw model text rather than dropping the segment.
                perSegment.append(oneResponse.content.trimmingCharacters(in: .whitespacesAndNewlines))
            }
        }
        return perSegment
    }

    /// Issues one translation request, mapping a transport failure to a typed
    /// `ChapterTranslationError`.
    private func send(
        prompt: String,
        config: ResolvedAIProviderConfig
    ) async throws -> AIResponse {
        let request = AIRequest(
            actionType: .translate,
            bookFingerprint: nil,
            locator: nil,
            contextText: "",
            userPrompt: prompt,
            targetLanguage: nil,
            promptVersion: promptVersion)
        do {
            return try await sender.sendTranslationRequest(request, using: config)
        } catch is CancellationError {
            throw ChapterTranslationError.cancelled
        } catch let error as ChapterTranslationError {
            throw error
        } catch {
            // Bug #320: keep the RAW detail (enum-case syntax + provider JSON) in
            // the log only; the thrown `.providerFailed` carries a sanitized,
            // user-facing message (`sanitizedProviderMessage`).
            log.error("Chapter translation failed: \(String(describing: error), privacy: .public)")
            throw Self.mapTransportError(error)
        }
    }

    /// Maps a raw transport error to a typed `ChapterTranslationError`,
    /// distinguishing a genuine offline failure (so the caller can render
    /// source-only per edge case (c)) from a generic provider failure.
    ///
    /// Only a `URLError` carrying an unambiguous connectivity code is treated
    /// as `.offline`. `AIError.networkError` is deliberately NOT mapped to
    /// `.offline` — it is a catch-all also thrown for invalid responses and
    /// misconfigured base URLs, so mapping it to `.offline` would mis-drive
    /// the source-only fallback on a provider/config fault (Gate-4 round-2).
    private static func mapTransportError(_ error: Error) -> ChapterTranslationError {
        if let urlError = error as? URLError {
            switch urlError.code {
            case .notConnectedToInternet, .networkConnectionLost,
                 .dataNotAllowed, .cannotConnectToHost, .timedOut:
                return .offline
            default:
                return .providerFailed(sanitizedProviderMessage(error))
            }
        }
        return .providerFailed(sanitizedProviderMessage(error))
    }

    /// Bug #320: the user-facing message for a non-offline translation failure.
    /// NEVER `String(describing: error)` — that stringifies an `AIError` to its
    /// enum-case syntax + raw provider JSON (e.g. `providerError("HTTP 400: {…}")`),
    /// which leaks internals and reads as a crash. Mirrors the sanitized path the
    /// Chat/Summarize/Translate tabs use: a known `AIError` → its `errorDescription`,
    /// anything else → `localizedDescription`. The raw detail is kept in the log
    /// only (logged at the catch site).
    static func sanitizedProviderMessage(_ error: Error) -> String {
        if let aiError = error as? AIError, let description = aiError.errorDescription {
            return description
        }
        return error.localizedDescription
    }
}

File: android/app/src/main/kotlin/com/vreader/app/bilingual/ChapterTranslationService.kt
// Purpose: feature #131 WI-3 — translates one chapter/document unit for Android
// bilingual reading (parity with iOS ChapterTranslationService.swift). The
// host-agnostic segment→chunk→translate→decode→per-segment-fallback→cache pipeline.
// A cache HIT returns the stored segments with zero AiClient calls; a MISS segments
// the source (paragraph granularity in v1), chunks it under a char budget, sends one
// one-shot AiClient.chat per chunk, strictly decodes the JSON string-array, falls
// back to one-request-per-segment on a decode failure, degrades a single failed chunk
// to source-only (and does NOT cache a partial result — a re-read retries the gap),
// and caches the recombined translation only on FULL success.
//
// Key decisions (mirroring iOS):
// - ChapterTranslationError is a sealed *value* type (WI-1), not a Throwable, so the
//   service raises [ChapterTranslationException] wrapping it — callers branch on
//   `.error`. `from()` is the AiError→ChapterTranslationError mapper.
// - Dual cancellation (iOS ChapterTranslationService.swift:359-364, M2): BOTH a native
//   CancellationException AND a typed ChapterTranslationException(Cancelled) thrown by
//   a chunk are re-raised as Cancelled BEFORE the generic degrade — a cancel aborts the
//   whole translation and is NEVER degraded-and-cached.
// - ensureActive() runs between chunks AND immediately before the Room write, so a
//   cancel right before the upsert leaves no row.
// - A partially-degraded result is NOT cached (iOS Bug #330); a cache-write failure
//   does NOT fail the translation (rule 50 §6). translatePreSegmented caches the
//   fully-successful result under the ENUMERATE's count (iOS Bug #343) so a
//   toggle/reopen restores via cachedTranslation(expectedSegmentCount) with zero
//   provider calls.
//
// @coordinates-with: ChapterTranslationStore.kt, ChapterSegmenter.kt,
//   TranslationChunker.kt, TranslationChunkContract.kt, ChapterTranslationError.kt,
//   com.vreader.app.ai.AiClient / AiRequest, AiProviderProfile,
//   iOS vreader/Services/AI/ChapterTranslationService.swift,
//   dev-docs/plans/20260710-feature-131-android-bilingual-interlinear.md (WI-3)
package com.vreader.app.bilingual

import com.vreader.app.ai.AiClient
import com.vreader.app.ai.AiMessage
import com.vreader.app.ai.AiProviderProfile
import com.vreader.app.ai.AiRole
import com.vreader.app.ai.AiRequest
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive

/** The outcome of translating one unit. */
data class ChapterTranslationResult(
    /** One translated segment per source segment, in order. */
    val segments: List<String>,
    /** True when served entirely from the disk cache (no provider call). */
    val fromCache: Boolean,
)

/**
 * The throwable carrier for the [ChapterTranslationError] value type — the sealed
 * error is not itself a Throwable, so the service raises this and callers read
 * [error]. [cause] carries the original throwable for logging when there was one.
 */
class ChapterTranslationException(
    val error: ChapterTranslationError,
    cause: Throwable? = null,
) : Exception(
    when (error) {
        is ChapterTranslationError.ProviderFailed -> error.message
        ChapterTranslationError.Offline -> "offline"
        ChapterTranslationError.TimedOut -> "timed out"
        ChapterTranslationError.Cancelled -> "cancelled"
    },
    cause,
)

/**
 * Translates one chapter/document unit into a target language with a provider-aware
 * disk cache. All AI is reached through the injected [aiClient] (the #118 seam; the
 * fake in tests). [promptVersion] is the cache-identity component — v1 is
 * `bilingual-v1|g=paragraph`.
 */
class ChapterTranslationService(
    private val aiClient: AiClient,
    private val store: ChapterTranslationStore,
    private val promptVersion: String,
    private val maxCharsPerChunk: Int = DEFAULT_MAX_CHARS_PER_CHUNK,
) {

    /**
     * A CACHE-ONLY lookup that needs NO provider (iOS Bug #306). Serves the canonical
     * row only when its stored `sourceParagraphCount` still matches the source
     * re-segmented into [granularity] segments — or when [acceptCountMismatch] is set
     * (self-healing consumers that carry the enumerate contract). Returns null on a
     * miss or a strict count mismatch.
     */
    suspend fun cachedTranslation(
        bookKey: String,
        unit: TranslationUnitId,
        sourceText: String,
        targetLanguage: String,
        granularity: TranslationGranularity = TranslationGranularity.paragraph,
        acceptCountMismatch: Boolean = false,
    ): ChapterTranslationResult? {
        val key = lookupKey(bookKey, unit, targetLanguage)
        val cached = store.translation(key) ?: return null
        val segments = segment(sourceText, granularity)
        if (cached.sourceParagraphCount != segments.size && !acceptCountMismatch) return null
        return ChapterTranslationResult(cached.translatedSegments, fromCache = true)
    }

    /**
     * The divergence-fallback restore (iOS Bug #343): serves the canonical row only
     * when its STORED `sourceParagraphCount` equals [expectedSegmentCount] (the
     * caller's own structure — e.g. a DOM enumerate's block count), so
     * blocks↔segments pair 1:1 by contract. Needs NO source text and NO provider → a
     * cache hit restores with zero provider calls.
     */
    suspend fun cachedTranslation(
        bookKey: String,
        unit: TranslationUnitId,
        expectedSegmentCount: Int,
        targetLanguage: String,
    ): ChapterTranslationResult? {
        val key = lookupKey(bookKey, unit, targetLanguage)
        val cached = store.translation(key) ?: return null
        if (cached.sourceParagraphCount != expectedSegmentCount) return null
        return ChapterTranslationResult(cached.translatedSegments, fromCache = true)
    }

    /**
     * Translates [sourceText] into [targetLanguage]. Serves from cache on a hit; on a
     * miss segments (paragraph granularity in v1) → chunks → per-chunk one-shot
     * requests → strict decode → per-segment fallback → per-chunk graceful degrade →
     * cache-write on FULL success only. Raises [ChapterTranslationException] on a
     * provider failure or cancellation. When [bypassCacheRead] is set the cache read
     * is skipped (an explicit re-translate) — the write still runs and replaces the
     * row in place.
     */
    suspend fun translate(
        bookKey: String,
        unit: TranslationUnitId,
        sourceText: String,
        targetLanguage: String,
        providerProfile: AiProviderProfile,
        granularity: TranslationGranularity = TranslationGranularity.paragraph,
        bypassCacheRead: Boolean = false,
    ): ChapterTranslationResult {
        val key = lookupKey(bookKey, unit, targetLanguage)
        // Segment FIRST — the count is needed to detect a stale cache row.
        val segments = segment(sourceText, granularity)

        if (!bypassCacheRead) {
            val cached = store.translation(key)
            if (cached != null) {
                if (cached.sourceParagraphCount == segments.size) {
                    return ChapterTranslationResult(cached.translatedSegments, fromCache = true)
                }
                // Stale (source changed): drop the row and re-translate. A delete
                // failure does not block re-translation — the later upsert refreshes
                // the same key regardless.
                runCatching { store.delete(key) }
            }
        }

        if (segments.isEmpty()) {
            return ChapterTranslationResult(emptyList(), fromCache = false)
        }

        val translated = translateSegments(segments, targetLanguage, providerProfile)

        // ensureActive() before the write: a cancel right before the upsert leaves no row.
        currentCoroutineContext().ensureActive()
        if (!translated.hadDegrade) {
            writeCache(bookKey, unit, targetLanguage, providerProfile, translated.segments, segments.size)
        }
        return ChapterTranslationResult(translated.segments, fromCache = false)
    }

    /**
     * Translates a PRE-SEGMENTED list of source segments directly (iOS Bug #268/#330/
     * #343), bypassing [ChapterSegmenter] — the caller's OWN enumerated block texts
     * pair 1:1 with the result by construction. Same per-chunk graceful degrade +
     * dual-cancellation contract as [translate]; on FULL success only, caches under
     * the canonical key with the ENUMERATE's count as `sourceParagraphCount`. A
     * partial degrade is NOT cached; a cache-write failure does not fail the
     * translation. Returns a list the same length as [segments].
     */
    suspend fun translatePreSegmented(
        bookKey: String,
        unit: TranslationUnitId,
        segments: List<String>,
        targetLanguage: String,
        providerProfile: AiProviderProfile,
    ): List<String> {
        if (segments.isEmpty()) return emptyList()

        val translated = translateSegments(segments, targetLanguage, providerProfile)

        currentCoroutineContext().ensureActive()
        if (!translated.hadDegrade) {
            writeCache(bookKey, unit, targetLanguage, providerProfile, translated.segments, segments.size)
        }
        return translated.segments
    }

    // ── internals ─────────────────────────────────────────────

    private data class ChunkedTranslation(val segments: List<String>, val hadDegrade: Boolean)

    /**
     * Chunks [segments], translates each chunk with the dual-cancellation + graceful
     * per-chunk degrade contract, and recombines in source order. Throws on
     * cancellation (native or typed) and on an all-chunks-failure. Returns the
     * recombined segments plus whether ANY chunk degraded (source-only) — a degraded
     * result is never cached by the callers.
     */
    private suspend fun translateSegments(
        segments: List<String>,
        targetLanguage: String,
        providerProfile: AiProviderProfile,
    ): ChunkedTranslation {
        val chunks = TranslationChunker.chunk(segments, maxCharsPerChunk)
        val translated = MutableList(segments.size) { "" }

        var anyChunkSucceeded = false
        var hadDegrade = false
        var lastChunkError: ChapterTranslationException? = null

        for (chunk in chunks) {
            // Between-chunk cancellation → typed Cancelled.
            try {
                currentCoroutineContext().ensureActive()
            } catch (e: CancellationException) {
                throw ChapterTranslationException(ChapterTranslationError.Cancelled, e)
            }

            val chunkSegments = chunk.map { segments[it] }
            try {
                val chunkResult = translateChunk(chunkSegments, targetLanguage, providerProfile)
                for ((offset, segmentIndex) in chunk.withIndex()) {
                    translated[segmentIndex] = chunkResult[offset]
                }
                anyChunkSucceeded = true
            } catch (e: CancellationException) {
                // Native coroutine cancellation → abort as Cancelled (BEFORE degrade).
                throw ChapterTranslationException(ChapterTranslationError.Cancelled, e)
            } catch (e: ChapterTranslationException) {
                // A typed Cancelled from a chunk must ABORT, not degrade (M2).
                if (e.error is ChapterTranslationError.Cancelled) throw e
                // Any other typed failure → degrade this chunk to source-only, continue.
                hadDegrade = true
                lastChunkError = e
            }
        }

        // If EVERY chunk failed (a genuine outage, not one over-budget chunk),
        // surface the error rather than returning an all-source-only result.
        if (!anyChunkSucceeded && lastChunkError != null) throw lastChunkError
        return ChunkedTranslation(translated, hadDegrade)
    }

    /**
     * Translates one chunk: a single whole-chunk request with a strict JSON-array
     * decode; on any decode / count mismatch, falls back to one request per segment.
     * A single over-budget segment is sub-split into ≤budget pieces and rejoined.
     */
    private suspend fun translateChunk(
        chunkSegments: List<String>,
        targetLanguage: String,
        providerProfile: AiProviderProfile,
    ): List<String> {
        // An oversized SINGLE segment (its own chunk, not sub-split by the chunker) is
        // sub-split into ≤budget pieces, translated piece by piece, and rejoined into
        // the one segment's translation.
        if (chunkSegments.size == 1) {
            val pieces = TranslationChunker.subSplit(chunkSegments[0], maxCharsPerChunk)
            if (pieces.size > 1) {
                val joined = StringBuilder()
                for (piece in pieces) {
                    currentCoroutineContext().ensureActive()
                    val response = send(TranslationChunkContract.userPrompt(listOf(piece), targetLanguage), providerProfile)
                    joined.append(decodeOrRaw(response, expectedCount = 1).single())
                }
                return listOf(joined.toString())
            }
        }

        val response = send(TranslationChunkContract.userPrompt(chunkSegments, targetLanguage), providerProfile)
        runCatching { TranslationChunkContract.decode(response, expectedCount = chunkSegments.size) }
            .getOrNull()
            ?.let { return it }

        // Decode failed → per-segment fallback (still under the same provider).
        val perSegment = ArrayList<String>(chunkSegments.size)
        for (segment in chunkSegments) {
            currentCoroutineContext().ensureActive()
            val oneResponse = send(TranslationChunkContract.userPrompt(listOf(segment), targetLanguage), providerProfile)
            perSegment.add(decodeOrRaw(oneResponse, expectedCount = 1).single())
        }
        return perSegment
    }

    /** Decodes to [expectedCount] strings, or falls back to the raw (trimmed) text. */
    private fun decodeOrRaw(raw: String, expectedCount: Int): List<String> =
        runCatching { TranslationChunkContract.decode(raw, expectedCount) }
            .getOrElse { List(expectedCount) { raw.trim() } }

    /**
     * Issues one one-shot request, mapping a transport failure to a typed
     * [ChapterTranslationException]. A native CancellationException propagates
     * (cooperative cancellation); a typed [ChapterTranslationException] is rethrown
     * as-is; any other throwable maps via [ChapterTranslationError.from].
     */
    private suspend fun send(prompt: String, providerProfile: AiProviderProfile): String {
        val request = AiRequest(
            model = providerProfile.model.ifBlank { providerProfile.kind.defaultModel },
            messages = listOf(AiMessage(AiRole.user, prompt)),
            temperature = providerProfile.temperature,
            maxTokens = providerProfile.maxTokens,
        )
        return try {
            aiClient.chat(request).text
        } catch (e: CancellationException) {
            throw e
        } catch (e: ChapterTranslationException) {
            throw e
        } catch (e: Throwable) {
            throw ChapterTranslationException(ChapterTranslationError.from(e), e)
        }
    }

    private fun segment(sourceText: String, granularity: TranslationGranularity): List<String> =
        when (granularity) {
            TranslationGranularity.paragraph -> ChapterSegmenter.paragraphs(sourceText)
            TranslationGranularity.sentence -> ChapterSegmenter.sentences(sourceText)
        }

    private fun lookupKey(bookKey: String, unit: TranslationUnitId, targetLanguage: String): String =
        CachedTranslation.lookupKey(bookKey, unit.storageKey, targetLanguage, promptVersion)

    /** Writes the canonical row. A store-write failure does not fail the translation. */
    private suspend fun writeCache(
        bookKey: String,
        unit: TranslationUnitId,
        targetLanguage: String,
        @Suppress("UNUSED_PARAMETER") providerProfile: AiProviderProfile,
        segments: List<String>,
        sourceParagraphCount: Int,
    ) {
        runCatching {
            store.upsert(
                CachedTranslation(
                    bookKey = bookKey,
                    unitStorageKey = unit.storageKey,
                    targetLanguage = targetLanguage,
                    promptVersion = promptVersion,
                    translatedSegments = segments,
                    sourceParagraphCount = sourceParagraphCount,
                    createdAt = System.currentTimeMillis(),
                ),
            )
        }
    }

    companion object {
        /**
         * The per-chunk character budget — conservative, well under any mainstream
         * provider's context window so a chunk + its prompt scaffold fit comfortably.
         */
        const val DEFAULT_MAX_CHARS_PER_CHUNK = 6000
    }
}
